### PR TITLE
Add `--[no-]validate` flag and `validate` option

### PR DESCRIPTION
**Pull request by Mouvedia**
> Which issue, if any, is this issue related to?

It was inspired by:

> […by it's very nature Stylelint runs should gravitate towards becoming a `noop`.](https://github.com/stylelint/stylelint/issues/6869#issue-1729397616)

> Is there anything in the PR that needs further explanation?

1.

Most of the time your config is valid. Hence we ought to provide a way to turn `validateOptions` into a noop.
i.e. if you know what you are doing then skipping the validation of the options should be possible
e.g. faster CI, vscode extension

The default is `true`.
i.e. it's not a disruptive change

2.

My very basic/crude perf test compared `main` and `flag-option-validate` using time and [bulma](https://cdn.jsdelivr.net/npm/bulma@1.0.2/css/bulma.min.css):

```console
$ time ./bin/stylelint.mjs bulma.min.css -f compact
```

```json
{
  "extends": [
    "stylelint-config-standard"
  ],
  "validate": false
}
```
After removing the lower and upper outliers, I was getting between 50 and 500 ms of gains on my old macbook air.
Thorough testing is not needed because the outcome should be obvious.

3.

Caching is another alternative.
The flag/option is easier to introduce.
i.e. overloading `cache` is riskier

4.

only review the second commit

### CODE REVIEW

**Comment by ybiquitous** [comment]
> …by it's very nature Stylelint runs should gravitate towards becoming a `noop`.

The quotation above seems to come from <https://github.com/stylelint/stylelint/issues/6869#issue-1729397616>. If so, the link is incorrect.

**Comment by Mouvedia** [comment]
> > …by it's very nature Stylelint runs should gravitate towards becoming a `noop`.
> 
> The quotation above seems to come from [#6869 (comment)](https://github.com/stylelint/stylelint/issues/6869#issue-1729397616). If so, the link is incorrect.

Fixed.

**Comment by ybiquitous** [comment]
FYI: My benchmark in MacBook Pro M1 with Node v20.17.0, using an non-minified version of [`bulma.css`](https://cdn.jsdelivr.net/npm/bulma@1.0.2/css/bulma.css):

`--validate`: (avg user time: 2.426s)

```sh-session
$ for i in $(seq 1 5); do time ./bin/stylelint.mjs -c tmp/config-validate.mjs tmp/bulma.css -f compact --no-cache --validate 2>/dev/null ; done
./bin/stylelint.mjs -c tmp/config-validate.mjs tmp/bulma.css -f compact   2>   2.43s user 0.11s system 198% cpu 1.274 total
./bin/stylelint.mjs -c tmp/config-validate.mjs tmp/bulma.css -f compact   2>   2.43s user 0.10s system 202% cpu 1.248 total
./bin/stylelint.mjs -c tmp/config-validate.mjs tmp/bulma.css -f compact   2>   2.45s user 0.10s system 200% cpu 1.270 total
./bin/stylelint.mjs -c tmp/config-validate.mjs tmp/bulma.css -f compact   2>   2.41s user 0.09s system 200% cpu 1.248 total
./bin/stylelint.mjs -c tmp/config-validate.mjs tmp/bulma.css -f compact   2>   2.41s user 0.10s system 199% cpu 1.258 total
```

`--no-validate`: (avg user time: 2.408s)

```sh-session
$ for i in $(seq 1 5); do time ./bin/stylelint.mjs -c tmp/config-validate.mjs tmp/bulma.css -f compact --no-cache --no-validate 2>/dev/null ; done
./bin/stylelint.mjs -c tmp/config-validate.mjs tmp/bulma.css -f compact   2>   2.38s user 0.10s system 200% cpu 1.236 total
./bin/stylelint.mjs -c tmp/config-validate.mjs tmp/bulma.css -f compact   2>   2.48s user 0.11s system 201% cpu 1.285 total
./bin/stylelint.mjs -c tmp/config-validate.mjs tmp/bulma.css -f compact   2>   2.37s user 0.10s system 200% cpu 1.230 total
./bin/stylelint.mjs -c tmp/config-validate.mjs tmp/bulma.css -f compact   2>   2.41s user 0.10s system 200% cpu 1.253 total
./bin/stylelint.mjs -c tmp/config-validate.mjs tmp/bulma.css -f compact   2>   2.40s user 0.09s system 199% cpu 1.246 total
```

`tmp/config-validate.mjs`:

```js
export default {
	extends: ['stylelint-config-standard'],
};
```

File: docs/user-guide/cli.md [hunk 1]
```diff
@@ -124,6 +124,10 @@ A filename to assign the input. [More info](options.md#codefilename).
 
 Accept stdin input even if it is empty.
 
+### `--validate, --no-validate`
+
+Force enable/disable the validation of the rules' options.
```
Comment (ybiquitous) [inline docs/user-guide/cli.md:129]: [suggestion] Let's add a link to `options.md`:

```suggestion
Force enable/disable the validation of the rules' options. [More info](options.md#validate).
```

File: docs/user-guide/options.md [hunk 2]
```diff
@@ -212,6 +212,12 @@ CLI flags: `--report-needless-disables, --rd`
 
 Report `stylelint-disable` comments that don't actually match any lints that need to be disabled.
 
+## `validate`
+
+CLI flags: [`--validate, --no-validate`](cli.md#--validate---no-validate)
+
+Force enable/disable the validation of the rules' options.
```
Comment (ybiquitous) [inline docs/user-guide/options.md:219]: [suggestion] It seems helpful to have a description for the default value, e.g.

```suggestion
Force enable/disable the validation of the rules' options. Default: `true`.
```

File: docs/user-guide/options.md [hunk 3]
```diff
@@ -212,6 +212,12 @@ CLI flags: `--report-needless-disables, --rd`
...
+
```
Comment (ybiquitous) [inline docs/user-guide/options.md:220]: [suggestion] This is not mandatory, but having an example might be helpful. E.g.

````md
For example, your CI might be faster by skipping the validation:

```js
config = {
  validate: process.env['CI'] !== 'true'
};
```
````

I'm unsure, but editors like VS Code might have a similar environment variable.

**Review by ybiquitous — COMMENTED** [review]
@Mouvedia Thanks, I'm in favor of this improvement. 👍🏼 

The implementation sounds good to me, but I've left some suggestions on the documentation. Can you have a look?

**Comment by Mouvedia** [comment]
> MacBook Pro M1 with Node v20.17.0

I was using Node v17.9.1 + a `findLastIndex` polyfill on a 1,4 GHz Intel Core i5.
It would make a real difference if you had an older computer.
It's consistent on my side and it's not surprising because not running code cannot be slower.

File: docs/user-guide/cli.md (hunk 1 above)
Comment (Mouvedia) [reply docs/user-guide/cli.md:129]: That would have been misleading.
i.e. it was the same exact description

Ill add the example first.

File: docs/user-guide/options.md (hunk 2 above)
Comment (Mouvedia) [reply docs/user-guide/options.md:219]: I would have advised to do the same for `formatter` but it's being changed in #7826.

File: docs/user-guide/options.md (hunk 3 above)
Comment (Mouvedia) [inline docs/user-guide/options.md:220]:
````suggestion
For example, your CI might be faster by skipping the validation if [`rules`](configure.md#rules) didn't change.
e.g.

```js
export default {
  // …
  validate: process.env['CI'] !== 'true'
};
```
````

File: .changeset/yellow-chairs-laugh.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+"stylelint": minor
+---
+
+Added: `validate` option to Node.js API and `--validate` CLI flag
```
[the remark this answers is not in the document]
Comment (Mouvedia) [reply .changeset/yellow-chairs-laugh.md:5]: All flags support `no-` prefixing so it's not necessary.
i.e. `--validate=0` is equivalent

**Comment by Mouvedia** [comment]
I am not sure if I will support such a change—it might be too disruptive—but we might later decide to accept #7608.
In that case, the flag would also accept strings.
e.g. `--validate=schema` vs `--validate=util` (alias `--validate`)

**Review by ybiquitous — APPROVED** [review]

**Comment by ybiquitous** [comment]
Thanks LGTM👍

**Comment by Mouvedia** [comment]
Looking at

https://github.com/stylelint/stylelint/blob/7ee28f4e266ecb2ac35f093e46606798634fa067/lib/lintSource.mjs#L73

it seems that `config` is never `undefined` once we reach

https://github.com/stylelint/stylelint/blob/7ee28f4e266ecb2ac35f093e46606798634fa067/lib/lintSource.mjs#L108

hence I suppose that

https://github.com/stylelint/stylelint/blob/86437547861e164028cb23b3feec965c7816de00/types/stylelint/index.d.ts#L169

is optional for facilitating tests.

**Pull request merged**